File: dotnet/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/CharacterDecorator.cs
namespace DecoratorDesignPattern.Example1;

public abstract class CharacterDecorator : ICharacter
{
    protected ICharacter _character;

    protected CharacterDecorator(ICharacter character)
    {
        _character = character ?? throw new ArgumentNullException(nameof(character));
    }

    public virtual string GetName() => _character.GetName();
    public virtual int GetHealth() => _character.GetHealth();
    public virtual int GetAttackPower() => _character.GetAttackPower();
    public virtual int GetDefense() => _character.GetDefense();
    public virtual int GetSpeed() => _character.GetSpeed();
    public virtual string GetDescription() => _character.GetDescription();
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/src/CharacterDecorator.ts
import { ICharacter } from './ICharacter';

export abstract class CharacterDecorator implements ICharacter {
  constructor(protected readonly character: ICharacter) {
    if (!character) throw new Error('character cannot be null');
  }

  getName(): string { return this.character.getName(); }
  getHealth(): number { return this.character.getHealth(); }
  getAttackPower(): number { return this.character.getAttackPower(); }
  getDefense(): number { return this.character.getDefense(); }
  getSpeed(): number { return this.character.getSpeed(); }
  getDescription(): string { return this.character.getDescription(); }
}
